### PR TITLE
feature/http-server

### PR DESCRIPTION
**Pull request by gustavosinbandera1**
this PR implement a htttp server in order to configure ESP32 WiFi features and read params from ESP32

### CODE REVIEW

File: components/Network/src/WiFi.cpp [hunk 1]
```diff
@@ -112,6 +123,16 @@ esp_err_t WiFi::init()
...
+
+    tcpip_adapter_dhcpc_stop(TCPIP_ADAPTER_IF_STA); // Don't run a DHCP client
+    tcpip_adapter_ip_info_t ipInfo;
+    inet_pton(AF_INET, DEVICE_IP, &ipInfo.ip);
+    inet_pton(AF_INET, DEVICE_GW, &ipInfo.gw);
+    inet_pton(AF_INET, DEVICE_NETMASK, &ipInfo.netmask);
+    tcpip_adapter_set_ip_info(TCPIP_ADAPTER_IF_STA, &ipInfo); 
+
+
+
```
Comment (jeandudey) [inline components/Network/src/WiFi.cpp]:
```suggestion
```

Why is this code here? This breaks the configuration for users. Not all users use the same configuration for STA, using DHCP **shouldn't** be disabled.

File: components/Network/src/WiFi.cpp [hunk 2]
```diff
@@ -22,6 +22,17 @@ namespace network {
...
+#include <lwip/sockets.h>
+#define DEVICE_IP          "192.168.1.100"
+
+// The Gateway address where we wish to send packets.
+// This will commonly be our access point.
+#define DEVICE_GW          "192.168.1.254"
+
+// The netmask specification.
+#define DEVICE_NETMASK "255.255.255.0"
+
```
Comment (jeandudey) [inline components/Network/src/WiFi.cpp]:
```suggestion
```

See the comment above, it **isn't** needed to use a static IP.

File: components/WS/src/HttpServerHandler.h [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+#include "HttpResponse.h"
+#include "WiFi.h"
+#include <Battery.h>
+#include <FuelGauge.h>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <sstream>
+#include <stdint.h>
+/**
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]:
```suggestion

/**
```

Separate comment from includes.

File: components/WS/src/HttpServerHandler.h [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+ * @brief Handle path matching for an incoming HTTP request.
+ */
+static const char* TAG1 = "HTTP_SERVER";
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]:
```suggestion
static const char* TAG = "HTTP_SERVER";
```

File: components/WS/src/HttpServerHandler.h [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+class HttpServerHandler
+{
+public:
+    static void configureWstAp(std::map<std::string, std::string>& params)
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]:
```suggestion
    static void configureStaAp(std::map<std::string, std::string>& params)
```

File: components/WS/src/HttpServerHandler.h [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                esp_err_to_name(err));
+        }
+
+        err = fuel_gauge.temperature(esc::TempMeasure::Internal, &temp);
+        if (err != ESP_OK) {
+            ESP_LOGE(TAG1, "Can't get internal temperature, err = %s",
+                esp_err_to_name(err));
+        }
+
+        ESP_LOGI("available memory ram--->>", "%d\n", heap_caps_get_free_size(MALLOC_CAP_DEFAULT));
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]:
```suggestion
```

Remove this.

File: components/WS/src/HttpServerHandler.h [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+                configureWst(params);
+            } else if (iterator->second == "ap") {
+                configureAp(params);
+            }
+        }
+    }
+
+    static void readDeviceInfoHandler(HttpRequest* pHttpRequest, HttpResponse* pHttpResponse)
+    {
+        ostringstream s1;
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]: Move this declaration to where it's used first.

Comment (jeandudey) [reply components/WS/src/HttpServerHandler.h]: Use `std::ostringstream`.

File: components/WS/src/HttpServerHandler.h [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            std::string field[2];
+            int j = 0;
+
+            while (std::getline(stream, field[j], ':')) {
+                j++;
+            }
+            json_map.insert(std::make_pair(field[0], field[1]));
+
+            i++;
+        }
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]: What if we use [cJSON](https://github.com/DaveGamble/cJSON) for this? It isn't that big

File: components/WS/src/HttpServerHandler.h [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+        //this reply need to be send before to stop wifi adapter but in this moment the server is not able to send  a reply
+        pHttpResponse->setStatus(HttpResponse::HTTP_STATUS_OK, "OK");
+        //return response to the remote client
+        pHttpResponse->sendData("{\"status\":\"OK\"}");
+
+        configureWstAp(json_map);
+    }
+
+    static void configureWst(std::map<std::string, std::string>& params)
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]:
```suggestion
    static void configureSta(std::map<std::string, std::string>& params)
```

File: components/WS/src/HttpServerHandler.h [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+        }
+    }
+
+    static void turnWstHandler(HttpRequest* pHttpRequest, HttpResponse* pHttpResponse)
+    {
+    }
+
+    static void turnWapHandler(HttpRequest* pHttpRequest, HttpResponse* pHttpResponse)
+    {
+    }
```
Comment (jeandudey) [inline components/WS/src/HttpServerHandler.h]: Remove if unused

File: main/main.cpp [hunk 11]
```diff
@@ -129,19 +133,25 @@ extern "C" void app_main()
...
     httpServer.addPathHandler(HttpRequest::HTTP_METHOD_GET, "/", webSocketHandler);
-    httpServer.start(80);
+    httpServer.addPathHandler(HttpRequest::HTTP_METHOD_GET, "/read-device-info", &HttpServerHandler::readDeviceInfoHandler);
+
+    //httpServer.addPathHandler(HttpRequest::HTTP_METHOD_GET, "/turn-off-wap", &HttpServerHandler::turnWapHandler);
+    //httpServer.addPathHandler(HttpRequest::HTTP_METHOD_GET, "/turn-on-wap", &HttpServerHandler::turnWapHandler);
+    //httpServer.addPathHandler(HttpRequest::HTTP_METHOD_POST, "/set-up-wap", &HttpServerHandler::setUpWstApHandler);
+
+    //httpServer.addPathHandler(HttpRequest::HTTP_METHOD_GET, "/turn-off-wst", &HttpServerHandler::turnWstHandler);
+    //httpServer.addPathHandler(HttpRequest::HTTP_METHOD_GET, "/turn-on-wst", &HttpServerHandler::turnWstHandler);
```
Comment (jeandudey) [inline main/main.cpp]: Remove if unused.

File: main/main.cpp [hunk 12]
```diff
@@ -129,19 +133,25 @@ extern "C" void app_main()
...
+    httpServer.addPathHandler(HttpRequest::HTTP_METHOD_POST, "/set-up-wst-ap", &HttpServerHandler::setUpWstApHandler);
+
+    httpServer.start(2565);
```
Comment (jeandudey) [inline main/main.cpp]: Add this port to `defaults.h`

File: main/main.cpp [hunk 13]
```diff
@@ -18,22 +18,25 @@
 #include <freertos/task.h>
 
 #include <BLEPreferences.h>
```
Comment (luisan00) [inline main/main.cpp]: Remove this include

**Comment by luisan00** [comment]
this was resolved by #103

**Pull request closed without merging**